apps/emqx_machine/src/emqx_machine_signal_handler.erl

```diff
@@ -0,0 +1,63 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% This module implements a gen_event handler which
+%% swap-in replaces the default one from OTP.
+%% The kill signal (sigterm) is captured so we can
+%% perform graceful shutdown.
+-module(emqx_machine_signal_handler).
+
+-export([start/0, init/1, format_status/2,
+         handle_event/2, handle_call/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-include_lib("emqx/include/logger.hrl").
+
+start() ->
+    ok = gen_event:swap_sup_handler(
+           erl_signal_server,
+           {erl_signal_handler, []},
+           {?MODULE, []}).
+
+init({[], _}) -> {ok, #{}}.
+
+handle_event(sigterm, State) ->
+    ?ULOG("Received terminate signal, shutting down now~n", []),
+    emqx_machine_terminator:graceful(),
```

Comment:
handle early kill sigterm.
```erlang
case emqx_machine:is_ready() of 
  true -> emqx_machine_terminator:graceful();
  false -> init:stop()
end.
```

Reply:
it's handled inside of `emqx_machine_terminator:graceful()`